File: snippets/java-app/src/main/java/com/example/snippets/java/MapActivity.java
/*
 * Copyright 2026 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.example.snippets.java;

import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;
import android.widget.Toast;
import androidx.activity.EdgeToEdge;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import com.example.snippets.common.R;
import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.MapView;
import com.google.android.gms.maps.OnMapReadyCallback;
import com.google.android.gms.maps.model.CameraPosition;
import com.google.android.gms.maps.model.LatLng;
import java.util.ArrayList;
import java.util.List;
import android.widget.FrameLayout;
import com.google.android.gms.maps.GoogleMapOptions;

public class MapActivity extends AppCompatActivity {

    public static final String EXTRA_SNIPPET_TITLE = "snippet_title";

    public MapView mapView;
    private String snippetTitle;
    private String groupTitle;
    public GoogleMap googleMap;
    private int currentIndex = -1;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        EdgeToEdge.enable(this);
        setContentView(R.layout.activity_map);
        if (getSupportActionBar() != null) {
            getSupportActionBar().setBackgroundDrawable(new android.graphics.drawable.ColorDrawable(android.graphics.Color.parseColor("#E76F00")));
        }
        getWindow().setStatusBarColor(android.graphics.Color.parseColor("#C05D00"));
        android.view.View mainView = findViewById(R.id.map_container);
        if (mainView != null) {
            androidx.core.view.ViewCompat.setOnApplyWindowInsetsListener(mainView, (v, insets) -> {
                androidx.core.graphics.Insets systemBars = insets.getInsets(androidx.core.view.WindowInsetsCompat.Type.systemBars());
                v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom);
                return insets;
            });
        }
        // MapView will be dynamically initialized and added to map_view_holder in runSnippet()
        try {
            android.content.pm.ApplicationInfo appInfo = getPackageManager().getApplicationInfo(
                    getPackageName(), android.content.pm.PackageManager.GET_META_DATA);
            if (appInfo.metaData != null) {
                String apiKey = appInfo.metaData.getString("com.google.android.geo.API_KEY");
                if (apiKey == null || apiKey.isEmpty() || apiKey.equals("DEFAULT_API_KEY") || apiKey.equals("YOUR_API_KEY") || !apiKey.startsWith("AIza")) {
                    Toast.makeText(this, "ERROR: Invalid Google Maps API Key configured in secrets.properties", Toast.LENGTH_LONG).show();
                    Log.e("MapActivity", "Invalid MAPS_API_KEY: '" + apiKey + "'");
                    finish();
                    return;
                }
            }
        } catch (Exception e) {
            Log.e("MapActivity", "Failed to verify API key metadata", e);
        }

        if (getSupportActionBar() != null) {
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        }

        findViewById(R.id.snapshot_button)
                .setOnClickListener(
                        v -> {
                            if (googleMap != null) {
                                CameraPosition cam = googleMap.getCameraPosition();
                                LatLng target = cam.target;
                                double rawBearing = cam.bearing;
                                double bearing = (rawBearing % 360.0 + 360.0) % 360.0;

                                String codeSnippet = String.format(
                                        "CameraPosition.builder()\n    .target(new LatLng(%.6f, %.6f))\n    .zoom(%.1ff)\n    .tilt(%.1ff)\n    .bearing(%.1ff)\n    .build()",
                                        target.latitude,
                                        target.longitude,
                                        cam.zoom,
                                        cam.tilt,
                                        bearing);

                                android.content.ClipboardManager cm = (android.content.ClipboardManager) getSystemService(android.content.Context.CLIPBOARD_SERVICE);
                                if (cm != null) {
                                    cm.setPrimaryClip(android.content.ClipData.newPlainText("CameraPosition", codeSnippet));
                                }
                                Toast.makeText(MapActivity.this, "📷 Camera Pose copied to clipboard!\n" + codeSnippet, Toast.LENGTH_LONG).show();
                                Log.d("MapActivity", "Camera Pose copied:\n" + codeSnippet);
                            } else {
                                Toast.makeText(MapActivity.this, "Map not initialized yet", Toast.LENGTH_SHORT).show();
                            }
                        });

        findViewById(R.id.reset_view_button).setOnClickListener(v -> runSnippet());

        groupTitle = getIntent().getStringExtra("group_title");
        snippetTitle = getIntent().getStringExtra(EXTRA_SNIPPET_TITLE);
        final List<SnippetItemInfo> snippetList = new ArrayList<>();
        for (SnippetGroupInfo group : SnippetRegistry.getSnippetGroups()) {
            snippetList.addAll(group.getItems());
        }
        for (int i = 0; i < snippetList.size(); i++) {
            SnippetItemInfo item = snippetList.get(i);
            if (item.getTitle().equals(snippetTitle)
                    && (groupTitle == null || item.getGroupTitle().equals(groupTitle))) {
                currentIndex = i;
                break;
            }
        }
        if (currentIndex == -1 && !snippetList.isEmpty()) {
            currentIndex = 0;
        }
        if (currentIndex >= 0 && currentIndex < snippetList.size()) {
            SnippetItemInfo item = snippetList.get(currentIndex);
            groupTitle = item.getGroupTitle();
            snippetTitle = item.getTitle();
        }

        findViewById(R.id.purpose_button)
                .setOnClickListener(
                        v -> {
                            String desc = "Demonstrates " + snippetTitle;
                            if (currentIndex >= 0 && currentIndex < snippetList.size()) {
                                SnippetItemInfo item = snippetList.get(currentIndex);
                                if (item.getDescription() != null && !item.getDescription().isEmpty()) {
                                    desc = item.getDescription();
                                }
                            }
                            new androidx.appcompat.app.AlertDialog.Builder(MapActivity.this)
                                    .setTitle("ℹ️ " + (snippetTitle != null ? snippetTitle : "Snippet Purpose"))
                                    .setMessage("🎯 Purpose & Point:\n" + desc + "\n\n👀 What you should see / do:\nInteract with the live map on your device screen to test this capability.")
                                    .setPositiveButton("Got It", null)
                                    .show();
                        });

        findViewById(R.id.previous_button)
                .setOnClickListener(
                        v -> {
                            if (snippetList.isEmpty()) return;
                            if (currentIndex > 0) {
                                currentIndex--;
                            } else {
                                currentIndex = snippetList.size() - 1;
                            }
                            SnippetItemInfo item = snippetList.get(currentIndex);
                            groupTitle = item.getGroupTitle();
                            snippetTitle = item.getTitle();
                            runSnippet();
                        });

        findViewById(R.id.next_button)
                .setOnClickListener(
                        v -> {
                            if (snippetList.isEmpty()) return;
                            if (currentIndex < snippetList.size() - 1) {
                                currentIndex++;
                            } else {
                                currentIndex = 0;
                            }
                            SnippetItemInfo item = snippetList.get(currentIndex);
                            groupTitle = item.getGroupTitle();
                            snippetTitle = item.getTitle();
                            runSnippet();
                        });

        runSnippet();
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull android.view.MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            android.content.Intent intent = new android.content.Intent(this, JavaSnippetsActivity.class);
            intent.addFlags(android.content.Intent.FLAG_ACTIVITY_CLEAR_TOP);
            startActivity(intent);
            finish();
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    protected void onNewIntent(@NonNull android.content.Intent intent) {
        super.onNewIntent(intent);
        setIntent(intent);
        groupTitle = intent.getStringExtra("group_title");
        snippetTitle = intent.getStringExtra(EXTRA_SNIPPET_TITLE);
        runSnippet();
    }

    private void recreateMapView(String mapId) {
        FrameLayout holder = findViewById(R.id.map_view_holder);
        if (holder == null) return;

        if (this.mapView != null) {
            this.mapView.onPause();
            this.mapView.onStop();
            this.mapView.onDestroy();
            holder.removeView(this.mapView);
            this.mapView = null;
            this.googleMap = null;
        }

        GoogleMapOptions options = new GoogleMapOptions();
        if (mapId != null && !mapId.isEmpty() && !mapId.equals("YOUR_MAP_ID")) {
            options.mapId(mapId);
            Log.d("MapActivity", "Recreating MapView with Map ID: " + mapId);
        } else {
            Log.d("MapActivity", "Recreating MapView with default options (No Map ID)");
        }

        this.mapView = new MapView(this, options);
        holder.addView(this.mapView, new FrameLayout.LayoutParams(
                FrameLayout.LayoutParams.MATCH_PARENT, FrameLayout.LayoutParams.MATCH_PARENT));

        // Sync lifecycle
        this.mapView.onCreate(null);
        this.mapView.onStart();
        this.mapView.onResume();
    }

    protected void runSnippet() {
        if (snippetTitle != null) {
            String key = groupTitle != null ? groupTitle + " - " + snippetTitle : snippetTitle;
            SnippetItemInfo snippet = SnippetRegistry.snippets.get(key);
            if (snippet == null) {
                for (SnippetGroupInfo g : SnippetRegistry.getSnippetGroups()) {
                    for (SnippetItemInfo item : g.getItems()) {
                        if (item.getTitle().equals(snippetTitle)) {
                            snippet = item;
                            break;
                        }
                    }
                }
            }
            final SnippetItemInfo finalSnippet = snippet;
            if (finalSnippet != null) {
                final List<SnippetItemInfo> allSnippets = new ArrayList<>();
                for (SnippetGroupInfo g : SnippetRegistry.getSnippetGroups()) {
                    allSnippets.addAll(g.getItems());
                }
                for (int i = 0; i < allSnippets.size(); i++) {
                    if (allSnippets.get(i).getTitle().equals(finalSnippet.getTitle()) &&
                        (groupTitle == null || allSnippets.get(i).getGroupTitle().equals(finalSnippet.getGroupTitle()))) {
                        currentIndex = i;
                        break;
                    }
                }

                if (getSupportActionBar() != null) {
                    getSupportActionBar().setTitle("☕ " + finalSnippet.getTitle());
                }
                android.view.View topBar = findViewById(androidx.appcompat.R.id.action_bar);
                if (topBar != null) {
                    topBar.setOnClickListener(v -> {
                        android.view.View pb = findViewById(R.id.purpose_button);
                        if (pb != null) pb.performClick();
                    });
                }
                
                // Aggressively recreate MapView
                recreateMapView(BuildConfig.MAP_ID);
                SnippetRegistry.clearTrackedItems();
                android.widget.LinearLayout controlsContainer = findViewById(R.id.custom_controls_container);
                if (controlsContainer != null) {
                    controlsContainer.removeAllViews();
                    controlsContainer.setVisibility(android.view.View.GONE);
                }

                this.mapView.getMapAsync(new OnMapReadyCallback() {
                    @Override
                    public void onMapReady(@NonNull GoogleMap map) {
                        MapActivity.this.googleMap = map;
                        TrackedMap.resetMapToDefaults(map);
                        try {
                            finalSnippet.getAction().execute(MapActivity.this, map);
                            map.setOnMapLoadedCallback(() -> {
                                if (mapView != null) mapView.setContentDescription("MapLoaded");
                            });
                            new Handler(Looper.getMainLooper()).postDelayed(() -> {
                                if (mapView != null) mapView.setContentDescription("MapLoaded");
                            }, 2000);
                        } catch (Exception e) {
                            Toast.makeText(MapActivity.this, "Error: " + e.getMessage(), Toast.LENGTH_LONG).show();
                            e.printStackTrace();
                        }
                    }
                });
            } else {
                Toast.makeText(
                                MapActivity.this,
                                "Snippet not found: " + snippetTitle,
                                Toast.LENGTH_LONG)
                        .show();
            }
        }
    }

    @Override
    protected void onStart() {
        super.onStart();
        if (mapView != null) mapView.onStart();
    }

    @Override
    protected void onResume() {
        super.onResume();
        if (mapView != null) mapView.onResume();
    }

    @Override
    protected void onPause() {
        if (mapView != null) mapView.onPause();
        super.onPause();
    }

    @Override
    protected void onStop() {
        if (mapView != null) mapView.onStop();
        super.onStop();
    }

    @Override
    protected void onDestroy() {
        if (mapView != null) mapView.onDestroy();
        super.onDestroy();
    }

    @Override
    public void onLowMemory() {
        super.onLowMemory();
        if (mapView != null) mapView.onLowMemory();
    }

    @Override
    protected void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        if (mapView != null) mapView.onSaveInstanceState(outState);
    }
}

File: snippets/kotlin-app/src/main/java/com/example/snippets/kotlin/MapActivity.kt
/*
 * Copyright 2026 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.example.snippets.kotlin

import android.os.Bundle
import android.util.Log
import android.widget.Button
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.ViewCompat
import androidx.core.view.WindowCompat
import androidx.core.view.WindowInsetsCompat
import androidx.lifecycle.lifecycleScope
import com.example.snippets.common.R
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.MapView
import com.google.android.gms.maps.OnMapReadyCallback
import com.google.android.gms.maps.model.LatLng
import android.widget.FrameLayout
import com.google.android.gms.maps.GoogleMapOptions
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlin.time.Duration.Companion.seconds

class MapActivity : AppCompatActivity() {

    companion object {
        const val EXTRA_SNIPPET_TITLE = "snippet_title"
    }

    var mapView: MapView? = null
    var googleMap: GoogleMap? = null
    var currentIndex: Int = -1

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        WindowCompat.setDecorFitsSystemWindows(window, false)
        setContentView(R.layout.activity_map)

        supportActionBar?.setBackgroundDrawable(android.graphics.drawable.ColorDrawable(android.graphics.Color.parseColor("#7F52FF")))
        window.statusBarColor = android.graphics.Color.parseColor("#5835C4")

        val mainView = findViewById<android.view.View>(R.id.map_container)
        if (mainView != null) {
            ViewCompat.setOnApplyWindowInsetsListener(mainView) { v, insets ->
                val systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars())
                v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom)
                insets
            }
        }

        // MapView will be dynamically initialized and added to map_view_holder in runSnippet()
        try {
            val appInfo = packageManager.getApplicationInfo(
                packageName,
                android.content.pm.PackageManager.GET_META_DATA,
            )
            val apiKey = appInfo.metaData?.getString("com.google.android.geo.API_KEY")
            if (apiKey.isNullOrEmpty() || apiKey == "DEFAULT_API_KEY" || apiKey == "YOUR_API_KEY" || !apiKey.startsWith("AIza")) {
                Toast.makeText(
                    this,
                    "ERROR: Invalid Google Maps API Key configured in secrets.properties",
                    Toast.LENGTH_LONG,
                ).show()
                Log.e("MapActivity", "Invalid MAPS_API_KEY: '$apiKey'")
                finish()
                return
            }
        } catch (e: Exception) {
            Log.e("MapActivity", "Failed to verify API key metadata", e)
        }

        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        val groupTitle = intent.getStringExtra("group_title")
        val snippetTitle = intent.getStringExtra(EXTRA_SNIPPET_TITLE)
        val snippetList = SnippetRegistry.getSnippetGroups().flatMap { it.items }
        currentIndex = snippetList.indexOfFirst {
            it.title == snippetTitle && (groupTitle == null || it.groupTitle == groupTitle)
        }
        Log.d("MapActivity", "onCreate: groupTitle='$groupTitle', snippetTitle='$snippetTitle', resolved index=$currentIndex, resolved item='${snippetList.getOrNull(currentIndex)?.let { "${it.groupTitle} - ${it.title}" }}'")
        if (currentIndex == -1 && snippetList.isNotEmpty()) {
            currentIndex = 0
            Log.d("MapActivity", "Fallback to index 0: '${snippetList[0].groupTitle} - ${snippetList[0].title}'")
        }

        findViewById<Button>(R.id.snapshot_button).apply {
            setOnClickListener {
                val map = googleMap
                if (map != null) {
                    val cam = map.cameraPosition
                    val center = cam.target
                    val rawBearing = cam.bearing
                    val bearing = (rawBearing % 360.0 + 360.0) % 360.0

                    val codeSnippet = """
                        CameraPosition.builder()
                            .target(LatLng(${center.latitude}, ${center.longitude}))
                            .zoom(${cam.zoom}f)
                            .tilt(${cam.tilt}f)
                            .bearing(${bearing}f)
                            .build()
                    """.trimIndent()

                    val cm = getSystemService(android.content.Context.CLIPBOARD_SERVICE) as? android.content.ClipboardManager
                    cm?.setPrimaryClip(android.content.ClipData.newPlainText("CameraPosition", codeSnippet))
                    Toast.makeText(this@MapActivity, "📷 Camera Pose copied to clipboard!\n$codeSnippet", Toast.LENGTH_LONG).show()
                    Log.d("MapActivity", "Camera Pose copied:\n$codeSnippet")
                } else {
                    Toast.makeText(this@MapActivity, "Map NOT initialized yet", Toast.LENGTH_SHORT).show()
                }
            }
        }

        findViewById<Button>(R.id.purpose_button).apply {
            setOnClickListener {
                val item = snippetList.getOrNull(currentIndex)
                val desc = if (!item?.description.isNullOrEmpty()) item?.description else "Demonstrates ${item?.title ?: snippetTitle}"
                androidx.appcompat.app.AlertDialog.Builder(this@MapActivity)
                    .setTitle("ℹ️ ${item?.title ?: snippetTitle ?: "Snippet Purpose"}")
                    .setMessage("🎯 Purpose & Point:\n$desc\n\n👀 What you should see / do:\nInteract with the live map on your device screen to test this capability.")
                    .setPositiveButton("Got It", null)
                    .show()
            }
        }

        findViewById<Button>(R.id.reset_view_button).apply {
            setOnClickListener {
                if (currentIndex >= 0) {
                    val item = snippetList[currentIndex]
                    runSnippet(item.groupTitle, item.title)
                }
            }
        }

        findViewById<Button>(R.id.previous_button).apply {
            setOnClickListener {
                if (snippetList.isEmpty()) return@setOnClickListener
                currentIndex = if (currentIndex > 0) currentIndex - 1 else snippetList.size - 1
                val item = snippetList[currentIndex]
                runSnippet(item.groupTitle, item.title)
            }
        }

        findViewById<Button>(R.id.next_button).apply {
            setOnClickListener {
                if (snippetList.isEmpty()) return@setOnClickListener
                currentIndex = if (currentIndex < snippetList.size - 1) currentIndex + 1 else 0
                val item = snippetList[currentIndex]
                runSnippet(item.groupTitle, item.title)
            }
        }

        if (currentIndex >= 0 && snippetList.isNotEmpty()) {
            val item = snippetList[currentIndex]
            runSnippet(item.groupTitle, item.title)
        }
    }

    override fun onOptionsItemSelected(item: android.view.MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            val intent = android.content.Intent(this, KotlinSnippetsActivity::class.java).apply {
                addFlags(android.content.Intent.FLAG_ACTIVITY_CLEAR_TOP)
            }
            startActivity(intent)
            finish()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onNewIntent(intent: android.content.Intent) {
        super.onNewIntent(intent)
        setIntent(intent)
        val groupTitle = intent.getStringExtra("group_title")
        val snippetTitle = intent.getStringExtra(EXTRA_SNIPPET_TITLE)
        runSnippet(groupTitle, snippetTitle)
    }

    override fun onStart() {
        super.onStart()
        mapView?.onStart()
    }

    override fun onResume() {
        super.onResume()
        mapView?.onResume()
    }

    override fun onPause() {
        mapView?.onPause()
        super.onPause()
    }

    override fun onStop() {
        mapView?.onStop()
        super.onStop()
    }

    override fun onDestroy() {
        mapView?.onDestroy()
        super.onDestroy()
    }

    override fun onLowMemory() {
        super.onLowMemory()
        mapView?.onLowMemory()
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        mapView?.onSaveInstanceState(outState)
    }

    private fun recreateMapView(mapId: String?) {
        val holder = findViewById<FrameLayout>(R.id.map_view_holder) ?: return

        mapView?.let {
            it.onPause()
            it.onStop()
            it.onDestroy()
            holder.removeView(it)
        }
        mapView = null
        googleMap = null

        val options = GoogleMapOptions()
        if (!mapId.isNullOrEmpty() && mapId != "YOUR_MAP_ID") {
            options.mapId(mapId)
            Log.d("MapActivity", "Recreating MapView with Map ID: $mapId")
        } else {
            Log.d("MapActivity", "Recreating MapView with default options (No Map ID)")
        }

        val newMapView = MapView(this, options)
        holder.addView(newMapView, FrameLayout.LayoutParams(
            FrameLayout.LayoutParams.MATCH_PARENT, FrameLayout.LayoutParams.MATCH_PARENT
        ))
        newMapView.onCreate(null)
        newMapView.onStart()
        newMapView.onResume()
        mapView = newMapView
    }

    private fun runSnippet(groupTitle: String?, snippetTitle: String?) {
        if (snippetTitle == null) return
        val key = if (groupTitle != null) "$groupTitle - $snippetTitle" else snippetTitle
        val snippet = SnippetRegistry.snippets[key]
            ?: SnippetRegistry.getSnippetGroups().flatMap { it.items }.find { it.title == snippetTitle }
            ?: return

        val allSnippets = SnippetRegistry.getSnippetGroups().flatMap { it.items }
        val idx = allSnippets.indexOfFirst { it.title == snippet.title && (groupTitle == null || it.groupTitle == groupTitle) }
        if (idx != -1) {
            currentIndex = idx
        }

        supportActionBar?.title = "🟣 ${snippet.title}"
        findViewById<android.view.View?>(androidx.appcompat.R.id.action_bar)?.setOnClickListener {
            findViewById<android.view.View?>(R.id.purpose_button)?.performClick()
        }

        recreateMapView(BuildConfig.MAP_ID)
        SnippetRegistry.clearTrackedItems()
        findViewById<android.widget.LinearLayout>(R.id.custom_controls_container)?.apply {
            removeAllViews()
            visibility = android.view.View.GONE
        }

        mapView?.getMapAsync { map ->
            this.googleMap = map
            TrackedMap.resetMapToDefaults(map)
            try {
                snippet.action(this@MapActivity, map, lifecycleScope)
                map.setOnMapLoadedCallback {
                    mapView?.contentDescription = "MapLoaded"
                }
                lifecycleScope.launch {
                    delay(2.seconds)
                    mapView?.contentDescription = "MapLoaded"
                }
            } catch (e: Exception) {
                Log.e("MapActivity", "Error running snippet: ${e.message}")
            }
        }
    }
}
